File: 394-decode-string/394-decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> countStack = new Stack();
        Stack<String> stringStack = new Stack();
        int count = 0;
        String res = "";
        
        for(int i = 0; i < s.length(); i++){
            if(Character.isDigit(s.charAt(i))){
                count = count * 10 +(s.charAt(i) -'0');
            }else if(s.charAt(i) == '['){
                countStack.push(count);
                stringStack.push(res);
                res = "";
                count = 0;
            }else if (s.charAt(i) == ']'){
                StringBuilder temp = new StringBuilder(stringStack.pop());
                for(int j = 0; j < countStack.peek(); j++){
                    temp.append(res);
                }
                countStack.pop();
                res = temp.toString();
            }else{
                res+=s.charAt(i);
            }
        }
        return res;
    }
}

File: 394-decode-string/394-decode-string.cpp
class Solution {
public:
    string result;
    string decodeString(string s) {
        string temp = "";
        string num;
        for(int i = 0; i < s.size(); i++){
            if(isdigit(s[i])){
                num+=s[i];
                s[i] = '@';
            }else if(isalpha(s[i])){
                result+=s[i];
            }else if(s[i] == '['){
                s[i] = '@';
                helper(stoi(num),s,i,temp);
                num = "";
            }
            result+=temp;
            temp = "";
        }
        return result;
    }
        void helper(int num, string& partial,int j,string& temp){
            string a = "";
            string num1 = "";
            for(int i = j; i < partial.size(); i++){
                if(isdigit(partial[i])){
                    num1+=partial[i];
                    partial[i] = '@';
                }
                if(partial[i] == ']'){
                    partial[i] = '@';
                    for(int i = 0; i < num; i++){
                        temp+=a;
                    }
                    return;
                }
                if(partial[i] == '['){
                    partial[i] = '@';
                    helper(stoi(num1),partial,i,temp);
                    a+=temp;
                    temp = "";
                    num1 = "";
                }
                if(partial[i] != '@'){
                    a+=partial[i];
                    partial[i] ='@';
                }
            }
        }
};